pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -76,6 +84,72 @@ func (d ChangeSetWithVersionedRs) Calculate() ([]Change, error) {
 	return allChanges, nil
 }
 
+func (d ChangeSetWithVersionedRs) checkAndUpdateResWithMaxDurationAnn(existingRsGrouped map[string][]ctlres.Resource,
+	existingRs versionedResources) error {
+
+	exNonVersionResMap := map[string]ctlres.Resource{}
+
+	for _, res := range existingRs.NonVersioned {
+		resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+		exNonVersionResMap[resKey] = res
+	}
+
+	for _, res := range d.newRs {
+		if val, found := res.Annotations()[maxDurationAnnKey]; found {
+			duration, err := time.ParseDuration(val)
+			if err != nil {
+				return fmt.Errorf("Unable to parse max-duration for resource %s: %s", res.Description(), err.Error())
+			}
+
+			resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+			exRes, found := exNonVersionResMap[resKey]
+			if !found {
+				if exResGroup, found := existingRsGrouped[resKey]; found {
+					exRes = exResGroup[len(exResGroup)-1]
+				}
+			}
+
+			if exRes != nil {
+				err = addLastRenewedTimeAnn(res, exRes, duration)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func addLastRenewedTimeAnn(res, exRes ctlres.Resource, duration time.Duration) error {
```

Comment:
This function would be moved out too, and get scoped to the same struct as the new one